File: backend/app/schemas/gamification.py
import uuid
from datetime import datetime
from decimal import Decimal
from pydantic import BaseModel, Field


class BadgeRead(BaseModel):
    id: uuid.UUID
    name: str
    description: str
    icon: str
    category: str
    criteria_type: str
    criteria_value: int

    model_config = {"from_attributes": True}


class UserBadgeRead(BaseModel):
    id: uuid.UUID
    badge: BadgeRead
    earned_at: datetime

    model_config = {"from_attributes": True}


class PointEntryRead(BaseModel):
    id: uuid.UUID
    points: int
    reason: str
    resource_id: uuid.UUID | None
    created_at: datetime

    model_config = {"from_attributes": True}


class StudentPoints(BaseModel):
    total_points: int
    badges: list[UserBadgeRead]
    recent_points: list[PointEntryRead]


class LeaderboardEntry(BaseModel):
    student_id: uuid.UUID
    student_name: str
    avatar_url: str | None
    total_points: int
    badge_count: int
    rank: int


class ActivityCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    description: str = Field(min_length=1, max_length=2000)
    points: int = Field(ge=1, le=1000, default=10)
    category: str = Field(min_length=1, max_length=50, default="general")
    course_id: uuid.UUID | None = None
    is_active: bool = True


class ActivityUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, min_length=1, max_length=2000)
    points: int | None = Field(default=None, ge=1, le=1000)
    category: str | None = Field(default=None, min_length=1, max_length=50)
    course_id: uuid.UUID | None = None
    is_active: bool | None = None


class ActivityRead(BaseModel):
    id: uuid.UUID
    title: str
    description: str
    points: int
    category: str
    course_id: uuid.UUID | None
    created_by: uuid.UUID
    is_active: bool
    created_at: datetime
    completed: bool = False
    completed_at: datetime | None = None
    completion_count: int = 0

    model_config = {"from_attributes": True}

File: frontend/src/types/gamification.ts
export interface Badge {
  id: string
  name: string
  description: string
  icon: string
  category: string
  criteria_type: string
  criteria_value: number
}

export interface UserBadge {
  id: string
  badge: Badge
  earned_at: string
}

export interface PointEntry {
  id: string
  points: number
  reason: string
  resource_id: string | null
  created_at: string
}

export interface StudentPoints {
  total_points: number
  badges: UserBadge[]
  recent_points: PointEntry[]
}

export interface LeaderboardEntry {
  student_id: string
  student_name: string
  avatar_url: string | null
  total_points: number
  badge_count: number
  rank: number
}

export interface UpcomingAssignment {
  quiz_id: string
  quiz_title: string
  course_id: string
  course_title: string
  due_at: string | null
  time_limit_min: number | null
  max_attempts: number
  attempts_used: number
  is_overdue: boolean
}

export interface AdminReport {
  total_students: number
  total_teachers: number
  total_courses: number
  total_parents: number
  active_users_30d: number
  avg_platform_grade: number | null
  avg_attendance_rate: number | null
  courses: CourseReport[]
}

export interface CourseReport {
  course_id: string
  course_title: string
  teacher_name: string
  enrolled_count: number
  avg_grade: number | null
  completion_rate: number | null
  avg_attendance_rate: number | null
}